File: backend/Tickets.Application/Common/Responses/ApiResponse.cs
namespace Tickets.Application.Common.Responses
{
    /// <summary>
    /// Respuesta estándar de la API
    /// </summary>
    public class ApiResponse<T>
    {
        /// <summary>
        /// Indica si la operación fue exitosa
        /// </summary>
        public bool Success { get; set; }

        /// <summary>
        /// Datos de la respuesta
        /// </summary>
        public T? Data { get; set; }

        /// <summary>
        /// Mensaje de la respuesta
        /// </summary>
        public string? Message { get; set; }

        /// <summary>
        /// Información de error (si existe)
        /// </summary>
        public ErrorDetails? Error { get; set; }

        /// <summary>
        /// Crea una respuesta exitosa
        /// </summary>
        public static ApiResponse<T> SuccessResponse(T data, string? message = null)
        {
            return new ApiResponse<T>
            {
                Success = true,
                Data = data,
                Message = message
            };
        }

        /// <summary>
        /// Crea una respuesta de error
        /// </summary>
        public static ApiResponse<T> ErrorResponse(string message, string? code = null, object? details = null)
        {
            return new ApiResponse<T>
            {
                Success = false,
                Error = new ErrorDetails
                {
                    Code = code,
                    Message = message,
                    Details = details
                }
            };
        }
    }

    /// <summary>
    /// Detalles de un error
    /// </summary>
    public class ErrorDetails
    {
        /// <summary>
        /// Código del error
        /// </summary>
        public string? Code { get; set; }

        /// <summary>
        /// Mensaje de error
        /// </summary>
        public string? Message { get; set; }

        /// <summary>
        /// Detalles adicionales del error
        /// </summary>
        public object? Details { get; set; }
    }

    /// <summary>
    /// Resultado paginado
    /// </summary>
    public class PagedResult<T>
    {
        /// <summary>
        /// Lista de elementos
        /// </summary>
        public List<T> Items { get; set; } = new();

        /// <summary>
        /// Página actual
        /// </summary>
        public int Page { get; set; }

        /// <summary>
        /// Tamaño de página
        /// </summary>
        public int PageSize { get; set; }

        /// <summary>
        /// Total de elementos
        /// </summary>
        public int TotalItems { get; set; }

        /// <summary>
        /// Total de páginas
        /// </summary>
        public int TotalPages => (int)Math.Ceiling(TotalItems / (double)PageSize);

        /// <summary>
        /// Indica si hay página anterior
        /// </summary>
        public bool HasPreviousPage => Page > 1;

        /// <summary>
        /// Indica si hay página siguiente
        /// </summary>
        public bool HasNextPage => Page < TotalPages;
    }
}

File: frontend/lib/core/api/api_response.dart
/// Respuesta estándar de la API
class ApiResponse<T> {
  final bool success;
  final T? data;
  final String? message;
  final ErrorDetails? error;

  ApiResponse({
    required this.success,
    this.data,
    this.message,
    this.error,
  });

  factory ApiResponse.fromJson(
    Map<String, dynamic> json,
    T Function(dynamic)? fromJsonT,
  ) {
    return ApiResponse<T>(
      success: json['success'] ?? false,
      data: json['data'] != null && fromJsonT != null
          ? fromJsonT(json['data'])
          : json['data'],
      message: json['message'],
      error: json['error'] != null
          ? ErrorDetails.fromJson(json['error'])
          : null,
    );
  }
}

/// Detalles del error
class ErrorDetails {
  final String? code;
  final String? message;
  final Map<String, dynamic>? details;

  ErrorDetails({
    this.code,
    this.message,
    this.details,
  });

  factory ErrorDetails.fromJson(Map<String, dynamic> json) {
    return ErrorDetails(
      code: json['code'],
      message: json['message'],
      details: json['details'],
    );
  }
}

/// Resultado paginado
class PagedResult<T> {
  final List<T> items;
  final int totalRecords;
  final int pageNumber;
  final int pageSize;
  final int totalPages;
  final bool hasPreviousPage;
  final bool hasNextPage;

  PagedResult({
    required this.items,
    required this.totalRecords,
    required this.pageNumber,
    required this.pageSize,
  })  : totalPages = (totalRecords / pageSize).ceil(),
        hasPreviousPage = pageNumber > 1,
        hasNextPage = pageNumber < (totalRecords / pageSize).ceil();

  factory PagedResult.fromJson(
    Map<String, dynamic> json,
    T Function(Map<String, dynamic>) fromJsonT,
  ) {
    return PagedResult<T>(
      items: (json['items'] as List)
          .map((item) => fromJsonT(item as Map<String, dynamic>))
          .toList(),
      totalRecords: json['totalRecords'] ?? json['totalItems'] ?? 0,
      pageNumber: json['pageNumber'] ?? json['currentPage'] ?? 1,
      pageSize: json['pageSize'] ?? 10,
    );
  }

  // Getter aliases para compatibilidad
  int get totalItems => totalRecords;
  bool get hasMore => hasNextPage;

  PagedResult<T> copyWith({
    List<T>? items,
    int? totalRecords,
    int? pageNumber,
    int? pageSize,
  }) {
    return PagedResult<T>(
      items: items ?? this.items,
      totalRecords: totalRecords ?? this.totalRecords,
      pageNumber: pageNumber ?? this.pageNumber,
      pageSize: pageSize ?? this.pageSize,
    );
  }
}
